core/src/createTextMaskInputElement.js

```diff
@@ -9,174 +9,184 @@ const strObject = 'object'
+      // If the `pipe` returns `false` or a string, the block below turns it into an object that the rest
+      // of the code can work with.
+      if (pipeResults === false) {
+        // If the `pipe` rejects `conformedValue`, we use the `previousConformedValue`, and set `rejected` to `true`.
+        pipeResults = {value: previousConformedValue, rejected: true}
+      } else if (isString(pipeResults)) {
+        pipeResults = {value: pipeResults}
       }
+    }
 
-      inputElement.value = inputElementValue // set the input value
-      safeSetSelection(inputElement, adjustedCaretPosition) // adjust caret position
+    // Before we proceed, we need to know which conformed value to use, the one returned by the pipe or the one
+    // returned by `conformToMask`.
+    const finalConformedValue = (piped) ? pipeResults.value : conformedValue
+
+    // After determining the conformed value, we will need to know where to set
+    // the caret position. `adjustCaretPosition` will tell us.
+    const adjustedCaretPosition = adjustCaretPosition({
+      previousConformedValue,
+      previousPlaceholder,
+      conformedValue: finalConformedValue,
+      placeholder,
+      rawValue: safeRawValue,
+      currentCaretPosition,
+      placeholderChar,
+      indexesOfPipedChars: pipeResults.indexesOfPipedChars,
+      caretTrapIndexes
+    })
+
+    // Text Mask sets the input value to an empty string when the condition below is set. It provides a better UX.
+    const inputValueShouldBeEmpty = finalConformedValue === placeholder && adjustedCaretPosition === 0
+    const emptyValue = showMask ? placeholder : emptyString
+    const inputElementValue = (inputValueShouldBeEmpty) ? emptyValue : finalConformedValue
+
+    state.previousConformedValue = inputElementValue // store value for access for next time
+    state.previousPlaceholder = placeholder
+
+    // In some cases, this `update` method will be repeatedly called with a raw value that has already been conformed
+    // and set to `inputElement.value`. The below check guards against needlessly readjusting the input state.
+    // See https://github.com/text-mask/text-mask/issues/231
+
+    if (inputElement.value === inputElementValue) {
+      return
     }
+
+    inputElement.value = inputElementValue // set the input value
+
+    safeSetSelection(inputElement, adjustedCaretPosition, recalculate) // adjust caret position
+  }
+
+  return {
+    state,
+    update
   }
 }
 
-function safeSetSelection(element, selectionPosition) {
+function safeSetSelection(element, selectionPosition, recalculate) {
   if (document.activeElement === element) {
-    if (isAndroid) {
-      defer(() => element.setSelectionRange(selectionPosition, selectionPosition, strNone), 0)
+    if (isAndroid && recalculate && false) {
+      defer(recalculate, 0)
     } else {
       element.setSelectionRange(selectionPosition, selectionPosition, strNone)
```

Comment:
Also, this should be called every time, even when recalculation will take process later.